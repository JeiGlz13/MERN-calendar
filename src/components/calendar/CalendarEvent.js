import React from 'react';

export const CalendarEvent = ({event}) => {

    const {titulo, user} = event;
    return (
        <div className='flex flex-col focus:outline-none outline-none text-xs sm:text-sm md:text-base' >
            <span>{titulo}</span>
            <strong>- {user.name} </strong>
        </div>
    )
}
