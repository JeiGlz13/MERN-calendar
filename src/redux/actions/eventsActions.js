import Swal from "sweetalert2";
import { fetchWithToken } from "../../helpers/fetch";
import { getEventsByUser } from "../../helpers/getEventsByUser";
import { prepareEvents } from "../../helpers/prepareEvents";
import { types } from "../types/types";

export const eventStartAddNew = (event) =>{
    return async (dispatch, getState) =>{
        const {uid, name} = getState().auth;

        try {
            const resp = await fetchWithToken('events', event, 'POST');
            const body = await resp.json();

            if (body.ok) {
                event.id = body.evento.id;
                event.user = {
                    _id: uid,
                    name: name
                }
                dispatch(eventAddNew(event))
            }
        } catch (error) {
            console.log(error);
        }
    }
}

const eventAddNew = (event) =>({
    type: types.eventAddNew,
    payload: event
});

export const eventSetActive = (event) =>({
    type: types.eventSetActive,
    payload: event
});

export const eventClearActive = () =>({
    type: types.eventClearActiveEvent
});

export const eventStartUpdate = (event) =>{
    return async (dispatch) =>{
        try {
            const resp = await fetchWithToken(`events/${event.id}`, event, 'PUT');
            const body = await resp.json();

            if (body.ok) {
                dispatch(eventUpdate(event));
            }else{
                Swal.fire('Error', body.msg ,'error');
            }
        } catch (error) {
            console.log(error);
        }
    }
} 

const eventUpdate= (event) =>({
    type: types.eventUpdate,
    payload: event
})

export const eventStartDelete = () =>{
    return async (dispatch, getState) =>{
        const {id} = getState().events.activeEvent;
        try {
            const resp = await fetchWithToken(`events/${id}`, {}, 'DELETE');
            const body = await resp.json();

            if (body.ok) {
                dispatch(eventDeleted());
                Swal.fire('Success', 'Evento eliminado con exito', 'success');
            }else{
                Swal.fire('Error', body.msg ,'error');
            }
        } catch (error) {
            console.log(error);
        }
    }
} 

const eventDeleted = () =>({type: types.eventDeleted});

export const eventSetStartEnd = (start, end) =>({
    type: types.eventSetStartEnd,
    payload: {
        eventStart: start,
        eventEnd: end
    }
});

export const eventClearEvents = () =>({type: types.eventClear});

export const eventClearStartEnd = () => ({type: types.eventClearStartEnd});

export const eventStartLoading = () =>{
    return async (dispatch, getState) =>{
        try {
                const {uid} = getState().auth;
               const resp = await fetchWithToken('events/');
               const body = await resp.json();

               const myEvents = getEventsByUser(body.eventos, uid);

               const eventos = prepareEvents(myEvents);


               dispatch(eventLoaded(eventos));


        } catch (error) {
            console.log(error)
        }
    }
}

const eventLoaded = (eventos) =>({
    type: types.eventLoaded,
    payload: eventos

})