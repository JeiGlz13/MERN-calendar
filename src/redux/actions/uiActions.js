import { types } from "../types/types";

export const setLoginForm = () =>({
    type: types.setLoginForm,
})

export const setRegisterForm = () =>({
    type: types.setRegisterForm,
})