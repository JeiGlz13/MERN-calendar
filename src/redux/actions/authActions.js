import Swal from "sweetalert2";
import { fetchWithoutToken, fetchWithToken } from "../../helpers/fetch";
import { types } from "../types/types";
import { eventClearActive, eventClearEvents } from "./eventsActions";

export const startLogin = (email, password) =>{
    return async (dispatch) =>{
        const resp = await fetchWithoutToken('auth', {email, password}, 'POST');
        const body = await resp.json();
        if (body.ok) {
            localStorage.setItem('token', body.token);
            localStorage.setItem('token-init-date', new Date().getTime());

            dispatch(login({
                uid: body.uid,
                name: body.name
            }))
        }else{
            Swal.fire('Error', body.msg ,'error');
        }
    }
}

export const startRegister = (name, email, password) =>{
        return async (dispatch) =>{
            const resp = await fetchWithoutToken('auth/new', {name, email, password}, 'POST');
            const body = await resp.json();
            if (body.ok) {
                localStorage.setItem('token', body.token);
                localStorage.setItem('token-init-date', new Date().getTime());
    
                dispatch(login({
                    uid: body.uid,
                    name: body.name
                }))
            }else{
                Swal.fire('Error', body.msg ,'error');
            }
        }
}

export const startChecking = () =>{
    return async (dispatch) =>{
        const resp = await fetchWithToken('auth/renew');
            const body = await resp.json();
            Swal.fire({
                title: 'Iniciando Sesion',
                text: 'Please Wait',
                allowOutsideClick: false,
                didOpen: ()=>{
                    Swal.showLoading();
                }
            });
            if (body.ok) {
                localStorage.setItem('token', body.token);
                localStorage.setItem('token-init-date', new Date().getTime());
    
                dispatch(login({
                    uid: body.uid,
                    name: body.name
                }))
            }else{
                dispatch(checkingFinish());
            }
            Swal.close();
    }
}

const checkingFinish = () =>({type: types.authCheckingFinish})

export const login = (user) => ({
    type: types.authLogin,
    payload: user
})

export const startLogout = () =>{
    return  (dispatch) =>{
        localStorage.clear();
        dispatch(logout());
        dispatch(eventClearEvents());
        dispatch(eventClearActive());
    }
}

export const logout = () =>({
    type: types.authLogout
})