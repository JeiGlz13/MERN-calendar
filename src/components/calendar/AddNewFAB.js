import React from 'react'
import { useDispatch } from 'react-redux'
import { openModal } from '../../redux/actions/modalActions';

export const AddNewFAB = () => {
    const dispatch = useDispatch();

    const setOpenModal = () =>{
        dispatch(openModal());
    }
    return (
        <button
        className= "absolute z-50 bottom-4 right-4 bg-indigo-500 text-white text-xl sm:text-2xl md:text-3xl h-10 w-10 sm:h-12 sm:w-12 md:h-16 md:w-16  rounded-full m-2"
        onClick={setOpenModal} >
            <i className='fas fa-plus'></i>
        </button>
    )
}
