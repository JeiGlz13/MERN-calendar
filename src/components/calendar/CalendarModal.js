import React, { useEffect, useState } from 'react';
import Modal from 'react-modal';

import DateTimePicker from 'react-datetime-picker';
import moment from 'moment';
import Swal from 'sweetalert2';
import { useDispatch, useSelector } from 'react-redux';
import { closeModal } from '../../redux/actions/modalActions'
import {eventAddNew, eventClearActive, eventClearStartEnd, eventUpdate} from '../../redux/actions/eventsActions';


Modal.setAppElement('#root');

const now = moment().minutes(0).seconds(0).add(1, "hours");
const nowPlus1 = now.clone().add(1, "hours");


export const CalendarModal = () => {
	const {modalOpen} = useSelector(state => state.modal);
	const {activeEvent, setStart, setEnd} = useSelector(state => state.events);

	let initEvent = {
		titulo: '',
		descripcion: '',
		start: now.toDate(),
		end: nowPlus1.toDate()
	}

	const dispatch = useDispatch();

	const [dateStart, setDateStart] = useState(now.toDate());
	const [dateEnd, setDateEnd] = useState(nowPlus1.toDate());

	const [formValues, setFormValues] = useState(initEvent);

	const {titulo, descripcion, start, end} = formValues;

	useEffect(() => {
		if(setStart && setEnd){
			initEvent = {
				...initEvent,
				start: setStart,
				end: setEnd
			}
			setDateStart(setStart);
			setDateEnd(setEnd);
		}else{
			const firstDate = now.toDate();
			const secondDate = nowPlus1.toDate();
			initEvent = {
				...initEvent,
				start: firstDate,
				end: secondDate
			}
				setDateStart(firstDate);
				setDateEnd(secondDate);
		}
	}, [setStart, setEnd])

	useEffect(() => {
		if (activeEvent) {
			setFormValues(activeEvent);
		}else{
			setFormValues(initEvent);
		}
		
	}, [activeEvent, setFormValues])


	const handleInputChange = ({target}) =>{
		setFormValues({
			...formValues,
			[target.name]: target.value
		})
	}

	
	const handleStartDate = (e) =>{
		setDateStart(e);
		setFormValues({
			...formValues,
			start: e
		});
	}

	const handleEndDate = (e) =>{
		setDateEnd(e);
		setFormValues({
			...formValues,
			end: e
		});
	}

	const validateForm = () =>{
		const firstDate = moment(start);
		const secondDate = moment(end);
		if(firstDate.isSameOrAfter(secondDate)){
			Swal.fire("Error", "La fecha de fin debe ser luego de la de inicio", "error");
			return false}
		if((titulo.trim().length < 2)||(descripcion.trim().length < 2)){
			Swal.fire("Error", "Rellene todos los campos", "error");
			return false}
		Swal.fire("Success", "Nueva nota añadida con éxito", "success");
		return true;
		
	}

	const saveNewEvent = (event) =>{
		dispatch( eventAddNew({
			...event,
			user: {
				id: new Date().getTime(),
				name: 'Jeisson'
			}
		}) );
	}

	const updateEvent = (evemt) =>{
		dispatch(eventUpdate(evemt));
	}

	const handleSubmitForm = (e) =>{
		e.preventDefault();
		if(validateForm() === true){
			//TODO: send to database

			if(activeEvent){
				updateEvent(formValues);
			}else{
				saveNewEvent(formValues);
			}
			setCloseModal();
		}
	}


	const setCloseModal = () =>{
       dispatch(closeModal());
	   setFormValues(initEvent);
	   dispatch(eventClearActive());
	   dispatch(eventClearStartEnd())
    }

    return (
        <Modal
        isOpen={modalOpen}
       // onAfterOpen={afterOpenModal}
        onRequestClose={setCloseModal}
        closeTimeoutMS={200}
        className="py-6 flex flex-col justify-center sm:py-12 outline-none focus:outline-none border-0"
        overlayClassName="modal-fondo"
      >
        
       
	<div className ="relative py-3 sm:max-w-xl sm:mx-auto font-Poppins">
		<div
			className ="absolute inset-0 bg-gradient-to-r from-indigo-300 to-indigo-600 shadow-lg transform -skew-y-3 sm:skew-y-0 -rotate-3 sm:-rotate-6 rounded-2xl sm:rounded-3xl">
		</div>
		<div className ="relative px-8 py-4 bg-white shadow-lg rounded-2xl sm:rounded-3xl sm:px-14 sm:py-8">
			<div className ="max-w-md mx-auto outline-none focus:outline-none border-0">
				<div>
					<h1 className ="text-lg sm:text-2xl font-semibold text-indigo-700 px-12 sm:px-24">
						{
							(activeEvent) ? ("Editar evento") : ("Nuevo Evento")
						}
					</h1>
				</div>
				<div className ="divide-y divide-gray-200">
					<div className ="py-4 sm:py-8 text-sm leading-6 space-y-4 text-gray-700 sm:text-lg sm:leading-7 flex flex-col justify-center">

                        <div className ="relative">
							<input autocomplete="off" id="titulo" name="titulo" type="text"
							value={titulo} onChange={handleInputChange}
							className ="peer placeholder-transparent h-10 w-full border-b-2 border-gray-300 text-gray-900 focus:outline-none focus:borer-rose-600" placeholder="Titulo" />
							<label for="titulo" className ="absolute left-0 -top-3.5 text-gray-800 text-sm peer-placeholder-shown:text-base peer-placeholder-shown:text-gray-440 peer-placeholder-shown:top-2 transition-all peer-focus:-top-3.5 peer-focus:text-indigo-600 peer-focus:text-sm">Titulo</label>
						</div>

						<div className ="relative">
							<label className ="absolute left-0 -top-3.5 text-gray-800 text-base">Fecha y hora inicio</label>

							<DateTimePicker
								onChange={handleStartDate}
								value={dateStart}
								className="mt-4"
							/>
						</div>

						<div className ="relative">
							<label className ="absolute left-0 -top-3.5 text-gray-800 text-base">Fecha y hora final</label>

							<DateTimePicker
								onChange={handleEndDate}
								value={dateEnd}
								minDate={dateStart}
								className="mt-4"
							/>
						</div>

                        <div className ="relative">
                            <textarea id="descripcion" name="descripcion" rows="5" 
							value={descripcion} onChange={handleInputChange}
							className ="resize-none peer placeholder-transparent w-full border-b-2 border-gray-300 text-gray-900 focus:outline-none focus:borer-rose-600" type = "text" placeholder="Descripcion">
                            </textarea>
							<label for="descripcion" className ="absolute left-0 -top-3.5 text-gray-800 text-sm peer-placeholder-shown:text-base peer-placeholder-shown:text-gray-440 peer-placeholder-shown:top-2 transition-all peer-focus:-top-3.5 peer-focus:text-indigo-600 peer-focus:text-sm">Descripcion</label>
						</div>
					
						<div className ="relative">
							<button className ="bg-indigo-500 text-white rounded-xl px-4 py-2"
							onClick={handleSubmitForm} >Guardar</button>
						</div>
					</div>
				</div>
			</div>
		</div>
	</div>
      </Modal>
    )
}
