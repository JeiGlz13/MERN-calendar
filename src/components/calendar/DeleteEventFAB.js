import React from 'react'
import { useDispatch } from 'react-redux'
import { eventStartDelete } from '../../redux/actions/eventsActions';

export const DeleteEventFAB = () => {
    const dispatch = useDispatch();

    const handleDelete = () =>{
        dispatch(eventStartDelete());
    }
    return (
        <button
        className='absolute z-50 bottom-4 left-4 bg-red-500 text-white text-xl sm:text-2xl md:text-3xl h-10 w-10 sm:h-12 sm:w-12 md:h-16 md:w-16 rounded-full m-2'
        onClick={handleDelete} >
            <i className='fas fa-trash'></i>
        </button>
    )
}
