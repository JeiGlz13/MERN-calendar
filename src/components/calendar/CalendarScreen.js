import React, { useState } from 'react'
import { Navbar } from '../ui/ui_navbar/Navbar';
import {Calendar, momentLocalizer} from 'react-big-calendar'
import moment from 'moment';
import 'moment/locale/es';

import 'react-big-calendar/lib/css/react-big-calendar.css';
import { messages } from '../../helpers/calendarMessagesEs';
import { CalendarEvent } from './CalendarEvent';
import { CalendarModal } from './CalendarModal';
import { useDispatch, useSelector } from 'react-redux';
import { openModal } from '../../redux/actions/modalActions';
import { eventClearActive, eventSetActive, eventSetStartEnd } from '../../redux/actions/eventsActions';
import { AddNewFAB } from './AddNewFAB';
import { DeleteEventFAB } from './DeleteEventFAB';

moment.locale('es');
const localizer = momentLocalizer(moment);

export const CalendarScreen = () => {
    const {events, activeEvent} = useSelector(state => state.events);
    const dispatch = useDispatch();
    const initialState = localStorage.getItem('lastView') || 'month';
    const [lastView, setLastView] = useState(initialState);

    const onDoubleClick = (e) =>{
        dispatch(openModal());
    }

    const onSelectEvent = (e) =>{
        dispatch(eventSetActive(e));
    }

    const onViewChange = (e) =>{
        setLastView(e);
        localStorage.setItem('lastView', e);
    }

    const eventStyleGetter = (event, start, end, isSelected) =>{
        const style = {
            backgroundColor: 'rgb(99 102 241)',
            borderRadius: '0px',
            opacity: 0.9,
            display: 'block',
            color: 'white'
        }

        return{
            style
        }
    }

    const onSelectSlot = (e) =>{
        dispatch(eventSetStartEnd(e.start, e.end))
        dispatch(eventClearActive());
        dispatch(openModal());
    }

    return (
        <div className='flex flex-col h-screen font-Poppins' >
            <Navbar/>

            <Calendar
                localizer={localizer}
                events={events}
                startAccessor="start"
                endAccessor="end"
                messages={messages}
                eventPropGetter={eventStyleGetter}
                onDoubleClickEvent={ onDoubleClick }
                onSelectEvent={onSelectEvent}
                onSelectSlot={onSelectSlot}
                selectable = {true}
                onView={onViewChange}
                view= {lastView}
                components={{
                    event: CalendarEvent
                }}
            />

            {
                (activeEvent)&&(
                    <DeleteEventFAB/>
                )
            }
            
            <AddNewFAB/>

            <CalendarModal/>
        </div>
    )
}
