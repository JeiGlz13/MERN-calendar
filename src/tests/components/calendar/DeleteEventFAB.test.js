import React from 'react';
import {mount} from 'enzyme';
import { Provider } from 'react-redux';
import '@testing-library/jest-dom';

import configureStore from 'redux-mock-store';
import thunk from "redux-thunk";
import { DeleteEventFAB } from '../../../components/calendar/DeleteEventFAB';
import { eventStartDelete } from '../../../redux/actions/eventsActions';

jest.mock('../../../redux/actions/eventsActions', () => ({
    eventStartDelete: jest.fn()
}))

const middleware = [thunk];
const mockStore = configureStore(middleware);

const initState = {};

const store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={store}>
        <DeleteEventFAB/>
    </Provider>
);

describe('Pruebas en DeleteEventFAB', () => {
    test('Debe mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();

    });
    
    test('Debe llamar el eventStartDelete', () => {
       wrapper.find('button').prop('onClick')();
       
       expect(eventStartDelete).toHaveBeenCalled();
    });
    
})
