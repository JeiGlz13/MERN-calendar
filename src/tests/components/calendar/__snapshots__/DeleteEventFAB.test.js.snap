// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en DeleteEventFAB Debe mostrarse correctamente 1`] = `
<button
  className="absolute z-50 bottom-4 left-4 bg-red-500 text-white text-xl sm:text-2xl md:text-3xl h-10 w-10 sm:h-12 sm:w-12 md:h-16 md:w-16 rounded-full m-2"
  onClick={[Function]}
>
  <i
    className="fas fa-trash"
  />
</button>
`;
